table.go

```diff
@@ -47,84 +61,73 @@ type RoutingTable struct {
 // NewRoutingTable creates a new routing table with a given bucketsize, local ID, and latency tolerance.
 func NewRoutingTable(bucketsize int, localID ID, latency time.Duration, m peerstore.Metrics) *RoutingTable {
 	rt := &RoutingTable{
-		Buckets:     []*Bucket{newBucket()},
-		bucketsize:  bucketsize,
-		local:       localID,
-		maxLatency:  latency,
-		metrics:     m,
-		PeerRemoved: func(peer.ID) {},
-		PeerAdded:   func(peer.ID) {},
+		Buckets:        []*Bucket{newBucket()},
+		bucketsize:     bucketsize,
+		local:          localID,
+		maxLatency:     latency,
+		metrics:        m,
+		cplRefreshedAt: make(map[uint]time.Time),
+		PeerRemoved:    func(peer.ID) {},
+		PeerAdded:      func(peer.ID) {},
 	}
 
 	return rt
 }
 
-// GetAllBuckets is safe to call as rt.Buckets is append-only
-// caller SHOULD NOT modify the returned slice
-func (rt *RoutingTable) GetAllBuckets() []*Bucket {
-	rt.tabLock.RLock()
-	defer rt.tabLock.RUnlock()
-	return rt.Buckets
-}
+// GetTrackedCplsForRefresh returns the Cpl's we are tracking for refresh.
+// Caller is free to modify the returned slice as it is a defensive copy.
+func (rt *RoutingTable) GetTrackedCplsForRefresh() []CplRefresh {
+	rt.cplRefreshLk.RLock()
+	defer rt.cplRefreshLk.RUnlock()
 
-// GenRandPeerID generates a random peerID in bucket=bucketID
-func (rt *RoutingTable) GenRandPeerID(bucketID int) peer.ID {
-	if bucketID < 0 {
-		panic(fmt.Sprintf("bucketID %d is not non-negative", bucketID))
-	}
-	rt.tabLock.RLock()
-	bucketLen := len(rt.Buckets)
-	rt.tabLock.RUnlock()
+	cpls := make([]CplRefresh, len(rt.cplRefreshedAt))
 
-	var targetCpl uint
-	if bucketID > (bucketLen - 1) {
-		targetCpl = uint(bucketLen) - 1
-	} else {
-		targetCpl = uint(bucketID)
+	i := 0
+	for c, t := range rt.cplRefreshedAt {
+		cpls[i] = CplRefresh{c, t}
```

Comment:
```go
cpls := make([]CplRefresh, 0, len(rt.cplRefreshedAt))
...

for ... {
  cpls = append(cpls, CplRefresh{c, t})
}
```

Make takes a type, length (optional), and capacity (optional). When building a slice with a known length, you can allocate with a known capacity to avoid having to re-allocate every time you run out of capacity.

See: https://blog.golang.org/go-slices-usage-and-internals